src/coreclr/jit/flowgraph.cpp

```diff
@@ -4619,6 +4621,181 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, Ar
+
+    const unsigned numBlocks = loop->NumLoopBlocks();
+    if (numBlocks == 0)
+    {
+        // This should never happen?
+        printf("NONE?");
+    }
+    else if (numBlocks == 1)
+    {
+        // Should be able to assume the only block is the header, but just in case, use the normal visitor.
+        loop->VisitLoopBlocksReversePostOrder([](BasicBlock* block) {
+            printf(FMT_BB " ", block->bbNum);
+            return BasicBlockVisit::Continue;
+        });
+    }
+    else
+    {
+        BasicBlock* lexicalTopBlock     = loop->GetLexicallyTopMostBlock();
+        BasicBlock* lexicalBottomBlock  = loop->GetLexicallyBottomMostBlock();
+        BasicBlock* lexicalEndIteration = lexicalBottomBlock->Next();
+
+        // Add all the loop blocks to a set, then check all the blocks in the identified lexical range.
+        // If there are non-loop blocks found, or if we don't find all the loop blocks in the lexical walk
+        // (meaning the bbNums might not be properly ordered), we fail.
+        loop->VisitLoopBlocks([&](BasicBlock* loopBlock) {
+            BitVecOps::AddElemD(&loopTraits, outputBlocks, loopBlock->bbNewPostorderNum);
+            return BasicBlockVisit::Continue;
+        });
+        bool lexicallyDense = true; // assume the best
+        for (BasicBlock* block = lexicalTopBlock; (block != nullptr) && (block != lexicalEndIteration);
+             block             = block->Next())
+        {
+            if (!loop->ContainsBlock(block))
+            {
+                lexicallyDense = false;
+                break;
+            }
+            BitVecOps::RemoveElemD(&loopTraits, outputBlocks, block->bbNewPostorderNum);
+        }
+        if (!BitVecOps::IsEmpty(&loopTraits, outputBlocks))
+        {
+            lexicallyDense = false;
+        }
+
+        if (lexicallyDense)
+        {
+            printf("[" FMT_BB ".." FMT_BB "]", lexicalTopBlock->bbNum, lexicalBottomBlock->bbNum);
+        }
+        else
+        {
+            bool firstMember = true;
+            loop->VisitLoopBlocksReversePostOrder([&firstMember](BasicBlock* block) {
+                printf("%s" FMT_BB, firstMember ? "" : ";", block->bbNum);
+                firstMember = false;
+                return BasicBlockVisit::Continue;
+            });
+        }
+    }
+
+    // Dump Entry Edges, Back Edges, Exit Edges
+
+    printf("\n  Entry: ");
+    bool first = true;
+    for (FlowEdge* const edge : loop->EntryEdges())
+    {
+        printf("%s" FMT_BB " -> " FMT_BB, first ? "" : "; ", edge->getSourceBlock()->bbNum, loop->GetHeader()->bbNum);
+        first = false;
+    }
+
+    printf("\n  Exit: ");
+    first = true;
+    for (FlowEdge* const edge : loop->ExitEdges())
+    {
+        BasicBlock* const exitBlock = edge->getSourceBlock();
+        printf("%s" FMT_BB " ->", first ? "" : "; ", exitBlock->bbNum);
+        exitBlock->VisitRegularSuccs(comp, [=](BasicBlock* succ) {
+            if (!loop->ContainsBlock(succ))
```

Comment:
```suggestion
            if (fgGetPredForBlock(succ, exitBlock) == edge)
```
Otherwise the output will look strange for a `BBJ_SWITCH` with multiple exits, I think.

We should probably store the target block as part of the exit edges (until `FlowEdge` begins storing them).

A small terminology nit -- the LLVM/GCC terminology of loops call the `exitBlock` here for "exiting" blocks, while the actual "exit" block is the one outside the loop (`succ` in this case).